packages/react-reconciler/src/ReactMemoCache.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {MemoCache} from 'shared/ReactTypes';
+import type {Dispatcher} from './ReactInternalTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentDispatcher} = ReactSharedInternals;
+
+export const MEMO_CACHE_UNSET_SENTINEL: symbol = Symbol.for(
+  'react.usememocache_sentinel',
+);
+
+/**
+ * Implementation of `useMemoCache()` return type
+ */
+export default class MemoCacheImpl implements MemoCache {
+  data: Array<any>;
+
+  constructor(data: Array<any>) {
+    this.data = data;
+  }
+
+  get size(): number {
+    return this.data.length;
+  }
+
+  has(index: number): boolean {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    return this.data[index] !== MEMO_CACHE_UNSET_SENTINEL;
+  }
+
+  get(index: number): any {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    const value = this.data[index];
+    if (value === MEMO_CACHE_UNSET_SENTINEL) {
+      throw new Error('Memo cache index accessed before being initialized');
+    }
+    return value;
+  }
+
+  set(index: number, value: any) {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    this.data[index] = value;
+  }
+
+  reset(): void {
```

Comment:
we shouldn't need this anymore when we're using React's built-in memo cache as that correctly resets on throws during render.